tools/deploy_firmware.py

```diff
@@ -0,0 +1,288 @@
+        wait: int, optional
+            The number of seconds to wait between attempts.
+
+        Returns
+        -------
+        bool
+            True if the device is reachable via SSH, false otherwise.
+        """
+        for _ in range(attempts):
+            try:
+                subprocess.check_output(["ssh", "{}@{}"
+                                         .format(self.username, self.name), "true"])
+                return True
+            except subprocess.CalledProcessError as e:
+                print(e.output)
+                print("Waiting for the device to be reachable")
+                time.sleep(wait)
+        return False
+
+    def sysupgrade(self):
+        """Upgrade the device using sysupgrade."""
+        raise NotImplementedError
+
+    def upgrade_uboot(self):
+        """Upgrade the device through uboot."""
+        raise NotImplementedError
+
+
+class NetgearRax40(PrplwrtDevice):
+    """A Netgear rax40.
+    At the moment, the rax40 can only be updated through uboot using a tftp server.
+    """
+
+    tftp_dir = "/srv/tftp"
+    """The root directory of the tftp server. The image will be copied there"""
+
+    uboot_prompt = "GRX500 #"
+    """The u-boot prompt on the target."""
+
+    def upgrade_uboot(self):
+        """Upgrade the device through u-boot.
+
+        It requires a running tftp server listenning on the IP the
+        target device will expect (see `ipaddr` in the uboot environment).
+        """
+        serial_path = "/dev/{}".format(self.name)
+        if not os.path.exists(serial_path):
+            raise ValueError("The serial device {} does not exist!\n".format(serial_path)
+                             + "Please make sure you have an appropriate udev rule for it.")
+        print("Copying image '{}' to '{}'".format(self.image, self.tftp_dir))
+        shutil.copy(os.path.join(self.artifacts_dir, self.image), self.tftp_dir)
+        print("Image copied to {}.".format(self.tftp_dir))
+        with serial.Serial(serial_path, self.baudrate) as ser:
+            print("Connecting to serial")
+            shell = pexpect.fdpexpect.fdspawn(ser, logfile=sys.stdout.buffer)
+            if not shell.isalive():
+                raise ValueError("Unable to connect to the serial device!")
+            print("Connected")
+            # The console might not be active yet:
+            shell.sendline("")
+            # make the shell prompt appear:
+            self.set_prompt(shell)
+            shell.expect(self.shell_prompt)
+            # reboot:
+            shell.sendline("reboot")
+            shell.expect(["Hit any key to stop autoboot:",
+                          pexpect.EOF, pexpect.TIMEOUT], timeout=120)
+            # stop autoboot:
+            shell.sendline("")
+            shell.expect(self.uboot_prompt)
+            # do the actual upgrade:
+            shell.sendline("run update_fullimage")
+            shell.expect(self.uboot_prompt, timeout=600)
```

Comment:
It doesn't take that long, but since in most cases it will anyway not wait for the timeout, I went with a "big enough" number so that we don't have to change it later.
600 seconds is probably a bit exaggerated though, I'll change it to 300.